src/auslib/services/releases.py

```diff
@@ -312,26 +320,31 @@ def update_release(name, blob, old_data_versions, when, changed_by, trans):
                     )
                     set_by_path(new_data_versions, path, {"sc_id": sc_id, "change_type": "update", "data_version": 1})
                 else:
-                    dbo.release_assets.update(
+                    future = dbo.release_assets.async_update(
                         where={"name": name, "path": str_path},
                         what={"data": new_assets},
                         old_data_version=old_data_version,
                         changed_by=changed_by,
                         transaction=trans,
                     )
+                    futures.append(future)
                     set_by_path(new_data_versions, path, old_data_version + 1)
         else:
             if when:
                 sc_id = dbo.release_assets.scheduled_changes.insert(
                     name=name, path=str_path, data=item, when=when, change_type="insert", changed_by=changed_by, transaction=trans,
                 )
+                futures.append(future)
                 set_by_path(new_data_versions, path, {"sc_id": sc_id, "change_type": "insert", "data_version": 1})
             else:
+                future = dbo.release_assets.async_insert(name=name, path=str_path, data=item, changed_by=changed_by, transaction=trans)
+                futures.append(future)
                 set_by_path(new_data_versions, path, 1)
-                dbo.release_assets.insert(name=name, path=str_path, data=item, changed_by=changed_by, transaction=trans)
 
     # Raises if there are errors
-    createBlob(new_blob).validate(current_product, app.config["WHITELISTED_DOMAINS"])
+    createBlob(full_blob).validate(current_product, app.config["WHITELISTED_DOMAINS"])
+
+    loop.run_until_complete(asyncio.gather(*futures))
```

Comment:
I think any exceptions raised in the futures will be accessible in each `future.exception()`. If you don't check those, we will continue on error. Even if the intention is to continue on error, we likely want to log the exception message, which would involve checking each `future.exception()` and logging.

We deal with this in scriptworker via the `raise_future_exceptions` and related functions: https://github.com/mozilla-releng/scriptworker/blob/0460bd266c0935ab1ed23e511e7d300f39bee2fa/src/scriptworker/utils.py#L378-L439

This raises the question of whether you've tested with any failure cases manually, and whether we should add automated unit tests that test failure case behaviors, if you haven't already.

Reply:
This is a good point. I'll figure out the best thing to do, and add some tests.